src/storage/memory_store.ts

```diff
@@ -0,0 +1,821 @@
+  }
+
+  /**
+   * Get the address info from its bip32 index.
+   * @param index bip32 address index to search for
+   * @async
+   * @returns {Promise<IAddressInfo | null>} The address info or null if not in storage
+   */
+  async getAddressAtIndex(index: number): Promise<IAddressInfo | null> {
+    const addr = this.addressIndexes.get(index);
+    if (addr === undefined) {
+      // We do not have this index loaded on storage, it should be generated instead
+      return null;
+    }
+    return this.addresses.get(addr) as IAddressInfo;
+  }
+
+  /**
+   * Save the address in storage
+   * @param {IAddressInfo} info Info on address to save
+   * @async
+   * @returns {Promise<void>}
+   */
+  async saveAddress(info: IAddressInfo): Promise<void> {
+    if (!info.base58) {
+      throw new Error('Invalid address');
+    }
+    if (this.addresses.has(info.base58)) {
+      throw new Error('Already have this address');
+    }
+
+    // Saving address info
+    this.addresses.set(info.base58, info);
+    this.addressIndexes.set(info.bip32AddressIndex, info.base58);
+
+    if (this.walletData.currentAddressIndex === -1) {
+      this.walletData.currentAddressIndex = info.bip32AddressIndex;
+    }
+
+    if (info.bip32AddressIndex > this.walletData.lastLoadedAddressIndex) {
+      this.walletData.lastLoadedAddressIndex = info.bip32AddressIndex;
+    }
+  }
+
+  /**
+   * Check that an address is in our storage.
+   * @param {string} base58 Address to check.
+   * @async
+   * @returns A promise that resolves to wheather the address is saved in storage or no.
+   */
+  async addressExists(base58: string): Promise<boolean> {
+    return this.addresses.has(base58);
+  }
+
+  /**
+   * Get the current address.
+   *
+   * @param {boolean | undefined} markAsUsed If we should set the next address as current
+   * @async
+   * @returns {Promise<string>} The address in base58 format
+   */
+  async getCurrentAddress(markAsUsed?: boolean): Promise<string> {
+    const addressInfo = await this.getAddressAtIndex(this.walletData.currentAddressIndex);
+    if (!addressInfo) {
+      throw new Error('Current address is not loaded');
+    }
+
+    if (markAsUsed) {
+      // Will move the address index only if we have not reached the gap limit
+      this.walletData.currentAddressIndex = Math.min(this.walletData.lastLoadedAddressIndex, this.walletData.currentAddressIndex + 1);
+    }
+    return addressInfo.base58;
+  }
+
+  /* TRANSACTIONS */
+
+  /**
+   * Iterate on the transaction history.
+   *
+   * @param {string|undefined} tokenUid Only yield txs with this token.
+   *
+   * @async
+   * @returns {AsyncGenerator<IHistoryTx>}
+   */
+  async *historyIter(tokenUid?: string | undefined): AsyncGenerator<IHistoryTx> {
+    for (const tx of this.history.values()) {
+      if (tokenUid === undefined) {
+        yield tx;
+        continue;
+      }
+
+      // If a tokenUid is passed, we only yield the transaction if it has the token in one of our addresses
```

Comment:
I think that this kind of pre-calculation should be done together with the reliable integration, since it allow us to make less unnecessary recalculations.
I did not find a method on the wallets that requires the history filtered by token often, so calculating this list when required seems more efficient for now.